server/src/main/java/org/elasticsearch/action/search/SearchResponse.java

```diff
@@ -604,28 +676,263 @@ public boolean equals(Object o) {
             return total == clusters.total && successful == clusters.successful && skipped == clusters.skipped;
         }
 
         @Override
         public int hashCode() {
+            /// MP TODO: not sure what to do about successful and skipped here, since they are calculated differently
+            /// MP TODO: between CCS MRT=true and other search types and are not immutable for CCS
             return Objects.hash(total, successful, skipped);
         }
 
         @Override
         public String toString() {
-            return "Clusters{total=" + total + ", successful=" + successful + ", skipped=" + skipped + '}';
-        }
-
-        public String toStringExtended() {
-            return Strings.format(
-                "Clusters{total=%d, successful=%d, skipped=%d, remote=%d, ccsMinimizeRoundtrips=%s}",
-                total,
-                successful,
-                skipped,
-                remoteClusters,
-                ccsMinimizeRoundtrips
-            );
+            return "Clusters{total=" + total + ", successful=" + getSuccessful() + ", skipped=" + getSkipped() + '}';
+        }
+
+        /**
+         * @return true if any underlying Cluster objects have PARTIAL, SKIPPED, FAILED or RUNNING status.
+         */
+        public boolean hasPartialResults() {
+            for (Cluster cluster : clusterInfo.values()) {
+                switch (cluster.getStatus()) {
+                    case PARTIAL:
+                    case SKIPPED:
+                    case FAILED:
+                    case RUNNING:
+                        return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Represents the search metadata about a particular cluster involved in a cross-cluster search.
+     * The Cluster object can represent both the local cluster and a remote cluster.
+     * For the local cluster, clusterAlias should be specified as RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY.
+     * Its XContent is put into the "details" section the "_clusters" entry in the SearchResponse.
+     * This is not an immutable class, since it needs to be updated as the search progresses
+     * (especially important for async CCS searches).
+     */
+    public static class Cluster implements ToXContentFragment, Writeable {
+        private final String clusterAlias;
+        private final String indexExpression; // original index expression from the user for this cluster
+        private volatile Status status;  // can be updated in different threads
+        private final List<ShardSearchFailure> failures;
+        private final AtomicInteger totalShards;
+        private final AtomicInteger successfulShards;
+        private final AtomicInteger skippedShards;
+        private final AtomicInteger failedShards;
+        private final AtomicLong took;  // search latency in millis for this cluster sub-search
+        private final AtomicBoolean timedOut;
```

Comment:
The thread safety model of this class does not use synchronization because it holds to a single-writer principle elsewhere in the code.

The Cluster object is updated by one ActionListener callback thread. No other thread should ever update a given Cluster object. (Each is updated by a separate single ActionListener). 

However, the Cluster objects do need to be read by queries (e.g., `GET _async_search/:id`), so all variables are set as volatile or Atomic (or final for `clusterAlias` and `indexExpression`) to allow getting latest state for status queries.

Note: the status access (e.g., `GET _async_search/:id`) is "best effort" since there is no synchronization. It is possible that while the status is being read for a status response, the Cluster is being actively updated by a callback thread. This is considered acceptable since that would only happen for "intermediate state". The final state when the search is completed will always be fully accurate.

Reply:
Also I think technically `took` and `timedOut` could just be volatile primitives, since they are always just set once and never incremented.

Reply:
Is it an option to create a new Cluster object at every change and keep the class immutable? I think that would be easier to reason about compared to all the mutable state it has now.

Reply:
For MRT=true, yes, because each cluster has only a single callback thread, so you have a single writer per Cluster.

But for MRT=false, the callback thread is at the granularity of a shard. So during the CanMatch (SearchShards) and during the Query/Fetch Phase you have N threads (N=number of shards per cluster) all potentially updating the same Cluster object with specific shard info. So either you need a shard thread safe Cluster object OR, if we want to use immutable objects, we need a shared per-Cluster lock so that each shard callback thread and safely increment the successful/skipped/failed counters and the failure list without a data race that loses information.

I think both are feasible. For the latter, I think we would need the shared Clusters object (which no longer can be immutable for CCS) to construct a set of lock objects, one per Cluster. Then the shard callback threads (MRT=false) would need to obtain that lock in order to create a new Cluster object with the updated information and then overwrite the existing and then release the lock.

If build such as locking system, we might want to then wire the MRT=true paths to use it as well, though there with a single writer per cluster it isn't necessary.

Let me know your view (or if there's another option I'm not thinking of) on how to proceed.

Reply:
In the latest push I changed the `SearchResponse.Cluster` to be fully immutable. The `SearchResponse.Clusters` object holds an unmodifiable Map<String, AtomicReference<Cluster>>, where the key is the clusterAlias.

For MRT=true (this PR), each CCSActionListener (one per cluster) gets a reference to the AtomicReference holding the Cluster it is searching. It will then do a CAS operation to update state with a new Cluster object. This isn't strictly necessary for the MRT=true paths, since there is a single writer to the Map per Cluster.

But it sets up the concurrency model for MRT=false, where the updates will be done per shard, not per Cluster. For that work (future ticket), each updater thread will update a Cluster object by retrieving the existing Cluster object, incrementing the appropriate counters and other state (e.g., status or timedOut or failure lists) and then doing a CAS back into the AtomicReference. If that fails, it will grab the new Cluster being held by the AtomicReference and re-attempt to increment/add data to a new Cluster and again try to CAS it into the AtomicReference.